### PR TITLE
show how to connect  energy from local controls

### LINKED ISSUE #25: wish list: energy offset
> Other things we need to consider are:
> * ability add/subtract energy offset instead of making people adjust lattice constants

_Originally posted by @ambarb in https://github.com/bluesky/hklpy/issues/17#issuecomment-591471108_

### LINKED ISSUE #26: wish list: get wavelength from beam line controls
> Other things we need to consider are:
> * ability to convert the RBV of beamline

> Also, convert energy units appropriately (soft x-ray beamlines at NSLS-II use eV, not keV). In other words, rely on EPICS UNITS to decide how to convert.
> 
> If we continue to require a manual setting of the hklpy calculation energy, then these things are transparent.  I am just anticipating something more sophisticated as I am not sure how much you want to change.

_Originally posted by @ambarb in https://github.com/bluesky/hklpy/issues/17#issuecomment-591471108_

### LINKED ISSUE #42: set calc engine energy even if locked = 0
@prjemian regarding https://github.com/bluesky/hklpy/pull/41#issuecomment-702981988 

This is basically what we do now, but nicer in that we reduce chance of a mistake.  I ask if we can make the method for both locked and unlocked the same so as to have more of a consistent user experience.  Maybe this is an enhancement?

_Originally posted by @ambarb in https://github.com/bluesky/hklpy/pull/41#issuecomment-702985366_

### PR DESCRIPTION
**Pull request by prjemian**
fixes #25, fixes #26, fixes #42

### CODE REVIEW

**Comment by prjemian** [comment]
@ambarb, @gfabbris: If you have time, can you look over the documentation added here (in the changes) and see if it answers the requests in #25 & #26.  Seems a bit wordy to me.

**Comment by ambarb** [comment]
@prjemian I can this afternoon.  Hope that is ok

File: doc/source/diffract.rst [hunk 1]
```diff
@@ -2,6 +2,263 @@
...
+:class:`hkl.diffract.E4CV`   Eulerian 4-circle, horizontal scattering plane
+:class:`hkl.diffract.E6C`    Eulerian 6-circle
+:class:`hkl.diffract.K4CV`   Kappa 4-circle, vertical scattering plane
+:class:`hkl.diffract.K6C`    Kappa 6-circle
+:class:`hkl.diffract.TwoC`   2-circle
+:class:`hkl.diffract.Zaxis`  Z-axis
+===========================  ==========================
+
+These special-use geometries provided by the
+**hkl-c++** library are also provided:
```
Comment (mrakitin) [inline doc/source/diffract.rst:27]: `provided` is used twice in this sentence. I think it can be phrased as:
```rst
These special-use geometries are also provided by the
**hkl-c++** library:
```

File: doc/source/diffract.rst [hunk 2]
```diff
@@ -2,6 +2,263 @@
...
+* :class:`hkl.diffract.SoleilSiriusKappa`
+* :class:`hkl.diffract.SoleilSiriusTurret`
+* :class:`hkl.diffract.SoleilSixs`
+* :class:`hkl.diffract.SoleilSixsMed1p2`
+* :class:`hkl.diffract.SoleilSixsMed2p2`
+
+In all cases, see the **hkl-c++** documentation
+for further information on these geometries.
+
+https://people.debian.org/~picca/hkl/hkl.html#org7ea41dd
```
Comment (mrakitin) [inline doc/source/diffract.rst:41]: Will it be rendered as a link? Maybe it can be done more explicitly.

File: doc/source/diffract.rst [hunk 3]
```diff
@@ -2,6 +2,263 @@
...
+        gamma = Component(SoftPositioner)
+        delta = Component(SoftPositioner)
+
+        def __init__(self, *args, **kwargs):
+            """
+            start the SoftPositioner objects with initial values
+            """
+            super().__init__(*args, **kwargs)
+            for axis in self.real_positioners:
+                axis.move(0)
```
Comment (mrakitin) [inline doc/source/diffract.rst:102]: For this code-block, it can be wrapped into https://www.sphinx-doc.org/en/1.5/markup/code.html.

File: doc/source/diffract.rst [hunk 4]
```diff
@@ -2,6 +2,263 @@
...
+
+    sim6c = SimulatedE6C('', name='sim6c')
+
+``k4cv`` : kappa 4-circle with EPICS motor PVs
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+To control a kappa diffractometer (in 4-circle geometry
+with vertical scattering plane)
+where the motor axes are provided
+by EPICS PVs, use `ophyd.EpicsMotor`. [#]_
```
Comment (mrakitin) [inline doc/source/diffract.rst:114]: In `vim`, I'd do <kbd>g</kbd>+<kbd>q</kbd> combination to aling this text into 80 symbols for more readability. This also applies to some code above. I understand the rendered code will be the same, but for better readability and consistency it would be great to have it wrapped into 80 symbols blocks.

I'm not sure if `vscode` has this functionality.

**Review by mrakitin — COMMENTED** [review]
Looks good. I left a few sphinx-related and style comments.

File: doc/source/diffract.rst (hunk 1 above)
Comment (prjemian) [reply doc/source/diffract.rst:27]: I'll use a footnote for this.  URL will be rendered in HTML with `<a href="...` code.

File: doc/source/diffract.rst (hunk 2 above)
Comment (prjemian) [reply doc/source/diffract.rst:41]: It *will* be rendered as a link.  Converting to a footnote per above.

File: doc/source/diffract.rst (hunk 4 above)
Comment (prjemian) [reply doc/source/diffract.rst:114]: VSCode has a [*rewrap*](https://marketplace.visualstudio.com/items?itemName=stkb.rewrap) extension.

Comment (prjemian) [reply doc/source/diffract.rst:114]: FYI: *Rewrap* and many more extensions are bundled with *VSCode Essentials*.

File: doc/source/diffract.rst (hunk 3 above)
Comment (prjemian) [reply doc/source/diffract.rst:107]: done

File: doc/source/diffract.rst (hunk 4 above)
Comment (prjemian) [reply doc/source/diffract.rst]: reformatted to 72-char lines

Comment (prjemian) [reply doc/source/diffract.rst]: except in explicit circumstances where long line is verbatim from console

Comment (mrakitin) [reply doc/source/diffract.rst]: Sounds good.

**Review by mrakitin — APPROVED** [review]
Thanks, Pete!

File: doc/source/diffract.rst [hunk 5]
```diff
@@ -2,6 +2,277 @@
...
+        tth = Component(EpicsMotor, "sky:m4")
+
+        energy = Component(EpicsSignal, "optics:energy")
+        energy_EGU = Component(EpicsSignal, "optics:energy.EGU")
+        energy_update_calc = Component(
+            EpicsSignal,
+            "optics:energy_locked")
+        energy_offset = Component(Signal, value=0)
+
+        def _energy_changed(self, value=None, **kwargs):
```
Comment (gfabbris) [inline doc/source/diffract.rst:224]: Actually, I don't quite understand when this callback is triggered. Shouldn't we subscribe it to the energy?

**Comment by gfabbris** [comment]
otherwise looks good to me.

File: doc/source/diffract.rst (hunk 5 above)
Comment (prjemian) [reply doc/source/diffract.rst:224]: It's in the base class:  https://github.com/bluesky/hklpy/blob/main/hkl/diffract.py#L137-L138

Comment (prjemian) [reply doc/source/diffract.rst:224]: https://github.com/bluesky/hklpy/blob/716b8ac65a714cbe3a2b8fdf7c84b8909dc45d88/hkl/diffract.py#L137-L138

**Comment by ambarb** [comment]
If `optics:energy_locked` is 0 (do not update calc engine), and someone wants to update the calc engine's energy manually, is there a way to do this with the `energy_offset` applied?  I don't think this is part of the commit so I will continue to look, but maybe this is something we missed the first time around.

Otherwise the wording of things is clear.  I particularly appreciate you explaining the sign convention you employed for `energy_offset`.  I looked for the math first, so maybe move that explanation before the code block???

It might be, especially for the simulation example, to have links on how to setup orientation matrix for convenience.

**Comment by prjemian** [comment]
@ambarb:  It is possible to update the calc engine's energy manually.  Need to do the various steps in the `_energy_changed()` method:

If you know the numbers:

    diffractometer.calc.energy = (energy_eV + offset_eV) * 0.001

If you use the signals (and know the signals are in *eV*:

    diffractometer.calc.energy = (diffractometer.energy.get()+ diffractometer.energy_offset.get()) * 0.001

...

I'll add this and your other suggestions.

**Comment by prjemian** [comment]
slightly easier way is:

1. `%mov diffractometer.energy_update_calc 1`
1. `diffractometer._energy_changed(diffractometer.energy.get())`
1. `%mov diffractometer.energy_update_calc 0`

**Comment by prjemian** [comment]
BUT, setting `diffractometer.calc.energy` should appear before `diffractometer.energy` since the former is the standard, the latter is a convenience.  That can be stated in the first example.

**Comment by ambarb** [comment]
@prjemian regarding https://github.com/bluesky/hklpy/pull/41#issuecomment-702981988 

This is basically what we do now, but nicer in that we reduce chance of a mistake.  I ask if we can make the method for both locked and unlocked the same so as to have more of a consistent user experience.  Maybe this is an enhancement?

**Comment by prjemian** [comment]
Can do.
